doc/flowvisor.8
.ds PN flowvisor

.TH flowvisor 8 "May 2008" "FlowVisor" "FlowVisor Manual"

.SH NAME
flowvisor \- OpenFlow network slicing daemon

.SH AUTHOR

Rob Sherwood <rob.sherwood@stanford.edu>

.SH SYNOPSIS
.B flowvisor
[-d \fIdebug\fR] [-h] [-l] [-p \fIport\fR] \fIconfig.xml\fR 

.SH DESCRIPTION
The
.B flowvisor
program is a daemon that logically slices an OpenFlow-enabled switch into multiple, logical switches each controlled by a different OpenFlow controller. 

For more details, see:

"Can the Production Network Be the Test-bed?" 
Rob Sherwood, Glen Gibb, Kok-Kiong Yap, Guido Appenzeller, Martin Casado, Nick McKeown, Guru Parulkar.
Symposium for Operating Systems Design and Implementation, 2010

.SH OPTIONS

.TP
-d \fIdebug\fR
Overrides debugging threshold in config.xml and sets it to \fIdebug\fR.  Does not modify config.xml.
Mostly used for manual debugging.  The string \fIdebug\fR must be one of MOBUG, DEBUG, NOTE, INFO, WARN, ALERT, CRIT,
and FATAL (in order of increasing priority/verbosity) similar to syslog(3).

.TP
-h
Prints the version number and usage information and exits.


.TP
-l
Logs to stderr instead of to syslog()

.TP
-p \fIport\fR
Overrides listenning port from config.xml and sets it to \fIport\fR.  Does not modify config.xml.
Mostly used for manual debugging.

.SH ENVIRONMENT

.TP 
FV_JVM_ARGS

The override the default arguments to the JVM with those in $FV_JVM_ARGS.
May be useful for increasing memory efficiency or playing with special
JVM optimizations, like fancy garbage collectors.

.TP
FV_DEBUG_PORT

Turns on remote debugging support for the JVM and sets the JVM to
listen on port $FV_DEBUG_PORT.  This is useful for attaching a
remote debugger, e.g., `jdb -attach fvserver.foo.com:$FV_DEBUG_PORT`.
\fINOTE\fR: there is no security in the JVM debugging protocol, so
by enabling this option you potentially open your running FlowVisor
to remote control/manipulation.  FV_DEBUG_PORT is specified as
\fI[local_address:]port\fR, where \fIlocal_address\fR defaults to "all
interfaces", so if you are interested in binding to a specific
interface, e.g., the loopback interface on port 12345, can set
FV_DEBUG_PORT=localhost:12345 .  See
\fIhttp://download.oracle.com/javase/6/docs/technotes/guides/jpda/conninv.html#Invocation\fR
for more information.



.SH "SEE ALSO"

.BR fvctl (1)
.BR fvconfig (1)


RELEASE-NOTES
VeRTIGO 1.0.0 : May 31 2012
    * First public release

Flowvisor 0.8.1 : July 22 2011
    * Better logging to find delays and process starvation
    * Changed default GC to be mark/sweep
    * Source documentation improvements
    * FLOWVISOR-120: disabled DNS lookup for stats_desc
    * FLOWVISOR-118: fixed FvSlicer memleak in XID translation
    * FLOWVISOR-117: bind fails for override IPv4 stacks
    * FLOWVISOR-116: move configfile read before logging

FlowVisor 0.8.0 : July 13 2011
    * fixed problem with FlowIDs exhausting
    * fixed memleaks that caused instability
    * fixed dynamic ports and topology discovery issues
    * fixed bugs in Vendor Stats and sendlen handling
    * updated developer documentation
    * added better support for remote debugging and JVM tweaks
    * fixed IPv6 dual-stack local port binding bug
    * changed superuser from 'root' to 'fvadmin'
    * lots of bugs fixed and sanity checks implemented

FlowVisor 0.7.2 : Mar 17 2011
    * added changeSlice fvctl/API command
    * fixed big bug in drop rules for disconnected controllers
    * fixed Null Pointer Exception race condition; stops crashing
    * added MRTG formatting scripts (thanks Chris Tengi!)

FlowVisor 0.7.1 : Feb 22 2011
    * diabled flow tracking by default
    * fixed bug in getSwitchFlowDB

FlowVisor 0.7 : 
    * fixed big I/O bug that caused switch flapping
    * added msg statistics, see `fvctl get{Slice,Switch}stats`
    * added flow tracking, see `fvctl get{Switch,SliceRewrite}FlowDB`
        to enable, run `fvctl setConfig flowvisor\!track_flows true`
    * lots of performance improvements
        * flow setup time ~5x improved
    * lots of small bugs fixed
    * password prompt no longer echos plaintext
    * better error handling in create slice and change flowspace

FlowVisor 0.6 - total rewrite (in Java!)
    * Single, unified flowspace config, not divided by slice
    * lldp is now rewritten with a trailer, not by rewritting the dst mac
    * OFError msgs now correctly encap the offending packet
    * NEW FlowVisor UserAPI: query and manipulate slice info


src/org/flowvisor/FlowVisor.java
package org.flowvisor;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.LinkedList;
import java.util.List;

import org.apache.xmlrpc.webserver.WebServer;

import org.flowvisor.api.APIServer;
import org.flowvisor.api.JettyServer;
import org.flowvisor.config.ConfigError;
import org.flowvisor.config.FVConfig;
import org.flowvisor.events.FVEventHandler;
import org.flowvisor.events.FVEventLoop;
import org.flowvisor.exceptions.UnhandledEvent;
import org.flowvisor.log.FVLog;
import org.flowvisor.log.LogLevel;
import org.flowvisor.log.StderrLogger;
import org.flowvisor.log.ThreadLogger;
import org.flowvisor.message.FVMessageFactory;
import org.flowvisor.ofswitch.OFSwitchAcceptor;
import org.flowvisor.ofswitch.TopologyController;
import org.flowvisor.vtopology.topology_configurator.VTConfigInterface;
import org.openflow.example.cli.Option;
import org.openflow.example.cli.Options;
import org.openflow.example.cli.ParseException;
import org.openflow.example.cli.SimpleCLI;

public class FlowVisor {
	// VENDOR EXTENSION ID
	public final static int FLOWVISOR_VENDOR_EXTENSION = 0x80000001;

	// VERSION
	public final static String FLOWVISOR_VERSION = "flowvisor-0.8.0";
	public final static String VERTIGO_VERSION = "vertigo-0.1.0";

	// Max slicename len ; used in LLDP for now; needs to be 1 byte
	public final static int MAX_SLICENAME_LEN = 255;

	/********/
	String configFile; 
	List<FVEventHandler> handlers;

	private int port;

	private WebServer apiServer;
	static FlowVisor instance;

	FVMessageFactory factory; 

	private static final Options options = Options.make(new Option[] {
			new Option("d", "debug", LogLevel.NOTE.toString(),
					"Override default logging threshold in config"),
			new Option("l", "logging", "Log to stderr instead of syslog"),
			new Option("p", "port", 0, "Override port from config"),
			new Option("h", "help", "Print help"),

	});

	public FlowVisor() {
		this.port = 0;
		this.handlers = new ArrayList<FVEventHandler>();
		this.factory = new FVMessageFactory();
	}

	/*
	 * Unregister this event handler from the system
	 */

	/**
	 * @return the configFile
	 */
	public String getConfigFile() {
		return configFile;
	}

	/**
	 * @param configFile
	 *            the configFile to set
	 */
	public void setConfigFile(String configFile) {
		this.configFile = configFile;
	}

	/**
	 * @return the port
	 */
	public int getPort() {
		return port;
	}

	/**
	 * @param port
	 *            the port to set
	 */
	public void setPort(int port) {
		this.port = port;
	}

	/**
	 * @return the factory
	 */
	public FVMessageFactory getFactory() {
		return factory;
	}

	/**
	 * @param factory
	 *            the factory to set
	 */
	public void setFactory(FVMessageFactory factory) {
		this.factory = factory;
	}

	public synchronized boolean unregisterHandler(FVEventHandler handler) {
		if (handlers.contains(handler)) {
			handlers.remove(handler);
			return true;
		}
		return false;
	}

	public void run() throws ConfigError, IOException, UnhandledEvent {
		FlowVisor.setInstance(this);

				// init polling loop
		FVLog.log(LogLevel.INFO, null, "initializing poll loop");
		FVEventLoop pollLoop = new FVEventLoop();

		if (port == 0)
			port = FVConfig.getInt(FVConfig.LISTEN_PORT);

		// init topology discovery, if configured for it
		if (TopologyController.isConfigured())
			handlers.add(TopologyController.spawn(pollLoop));

		// init switchAcceptor
		OFSwitchAcceptor acceptor = new OFSwitchAcceptor(pollLoop, port, 16);
		handlers.add(acceptor);
		// start XMLRPC UserAPI server; FIXME not async!
		try {
			this.apiServer = APIServer.spawn();
		} catch (Exception e) {
			FVLog.log(LogLevel.FATAL, null, "failed to spawn APIServer");
			e.printStackTrace();
			System.exit(-1);
		}

		// print some system state
		boolean flowdb = false;
		try {
			if (FVConfig.getBoolean(FVConfig.FLOW_TRACKING))
				flowdb = true;
		} catch (ConfigError e) {
			// assume off if not set
			FVConfig.setBoolean(FVConfig.FLOW_TRACKING, false);
			this.checkPointConfig();
		}
		if (!flowdb)
			FVLog.log(LogLevel.INFO, null, "flowdb: Disabled");

		/**
		 * @authors roberto.doriguzzi matteo.gerola
		 */
		VTConfigInterface vt_config = new VTConfigInterface();
		vt_config.InitDB();
		
		// start event processing
		pollLoop.doEventLoop();
	}

	/**
	 * FlowVisor Daemon Executable Main
	 *
	 * Takes a config file as only parameter
	 *
	 * @param args
	 *            config file
	 * @throws Throwable
	 */

	public static void main(String args[]) throws Throwable {

		ThreadLogger threadLogger = new ThreadLogger();
		Thread.setDefaultUncaughtExceptionHandler(threadLogger);
		long lastRestart = System.currentTimeMillis();
		while (true) {
			FlowVisor fv = new FlowVisor();
			fv.parseArgs(args);

			try {
				// load config from file
				FVConfig.readFromFile(fv.configFile);

				JettyServer.spawnJettyServer();
				fv.run();
			} catch (Throwable e) {
				FVLog.log(LogLevel.CRIT, null, "MAIN THREAD DIED!!!");
				FVLog.log(LogLevel.CRIT, null, "----------------------------");
				threadLogger.uncaughtException(Thread.currentThread(), e);
				FVLog.log(LogLevel.CRIT, null, "----------------------------");
				if ((lastRestart + 5000) > System.currentTimeMillis()) {
					System.err.println("respawning too fast -- DYING");
					FVLog.log(LogLevel.CRIT, null,
							"respawning too fast -- DYING");
					fv.tearDown();
					throw e;
				} else {
					FVLog.log(LogLevel.CRIT, null,
							"restarting after main thread died");
					lastRestart = System.currentTimeMillis();
					fv.tearDown();
				}
				fv = null;
				System.gc(); // give the system a bit to clean up after itself
				Thread.sleep(1000);
			}
		}
	}

	private void parseArgs(String[] args) {
		SimpleCLI cmd = null;
		try {
			cmd = SimpleCLI.parse(options, args);

		} catch (ParseException e) {
			usage("ParseException: " + e.toString());
		}
		if (cmd == null)
			usage("need to specify arguments");
		int i = cmd.getOptind();
		if (i >= args.length)
			usage("need to specify a configuration");
		setConfigFile(args[i]);

		if (cmd.hasOption("d")) {
			FVLog.setThreshold(LogLevel.valueOf(cmd.getOptionValue("d")));
			System.err.println("Set default logging threshold to "
					+ FVLog.getThreshold());
		}
		if (cmd.hasOption("l")) {
			System.err.println("Setting debugging mode: all logs to stderr");
			FVLog.setDefaultLogger(new StderrLogger());
		}
		if (cmd.hasOption("p")) {
			int p = Integer.valueOf(cmd.getOptionValue("p"));
			setPort(p);
			System.err.println("Overriding port from config: setting to "
					+ getPort());
		}

	}

	private void tearDown() {
		if (this.apiServer != null)
			this.apiServer.shutdown(); // shutdown the API Server
		List<FVEventHandler> tmp = this.handlers;
		this.handlers = new LinkedList<FVEventHandler>();
		for (Iterator<FVEventHandler> it = tmp.iterator(); it.hasNext();) {
			FVEventHandler handler = it.next();
			it.remove();
			handler.tearDown();
		}
	}

	/**
	 * Print usage message and warning string then exit
	 *
	 * @param string
	 *            warning
	 */

	private static void usage(String string) {
		System.err.println("VeRTIGO version: " + VERTIGO_VERSION);
		System.err.println("Roberto Doriguzzi: roberto.doriguzzi@create-net.org");
		System.err.println("Matteo Gerola: matteo.gerola@create-net.org");
		System.err.println("");
		System.err.println("FlowVisor version: " + FLOWVISOR_VERSION);
		System.err.println("Rob Sherwood: rsherwood@telekom.com/rob.sherwood@stanford.edu");
		System.err.println("---------------------------------------------------------------");
		System.err.println("err: " + string);
		SimpleCLI.printHelp("vertigo [options] config.xml",
				FlowVisor.getOptions());
		System.exit(-1);
	}

	private static Options getOptions() {

		return FlowVisor.options;
	}

	/**
	 * Get the running fv instance
	 *
	 * @return
	 */
	public static FlowVisor getInstance() {
		return instance;
	}

	/**
	 * Set the running fv instance
	 *
	 * @param instance
	 */
	public static void setInstance(FlowVisor instance) {
		FlowVisor.instance = instance;
	}

	/**
	 * Returns a unique, shallow copy of the list of event handlers registered
	 * in the flowvisor
	 *
	 * Is unique to prevent concurrency problems, i.e., when wakling through the
	 * list and a handler gets deleted
	 *
	 * @return
	 */
	public synchronized ArrayList<FVEventHandler> getHandlersCopy() {
		return new ArrayList<FVEventHandler>(handlers);
	}

	public void addHandler(FVEventHandler handler) {
		this.handlers.add(handler);
	}

	public void removeHandler(FVEventHandler handler) {
		this.handlers.remove(handler);
	}

	public void setHandlers(ArrayList<FVEventHandler> handlers) {
		this.handlers = handlers;
	}

	/**
	 * Save the running config back to disk
	 *
	 * Write to a temp file and only if it succeeds, move it into place
	 *
	 * FIXME: add versioning
	 */
	public void checkPointConfig() {
		String tmpFile = this.configFile + ".tmp"; // assumes no one else can
		// write to same dir
		// else security problem

		// do we want checkpointing?
		try {
			if (!FVConfig.getBoolean(FVConfig.CHECKPOINTING))
				return;
		} catch (ConfigError e1) {
			FVLog.log(LogLevel.WARN, null,
					"Checkpointing config not set: assuming you want checkpointing");
		}

		try {
			FVConfig.writeToFile(tmpFile);
		} catch (FileNotFoundException e) {
			FVLog.log(LogLevel.CRIT, null,
					"failed to save config: tried to write to '" + tmpFile
							+ "' but got FileNotFoundException");
			return;
		}
		// sometimes, Java has the stoopidest ways of doing things :-(
		File tmp = new File(tmpFile);
		if (tmp.length() == 0) {
			FVLog.log(LogLevel.CRIT, null,
					"failed to save config: tried to write to '" + tmpFile
							+ "' but wrote empty file");
			return;
		}

		tmp.renameTo(new File(this.configFile));
		FVLog.log(LogLevel.INFO, null, "Saved config to disk at "
				+ this.configFile);
	}

	public String getInstanceName() {
		// TODO pull from FVConfig; needed for slice stiching
		return "magic flowvisor1";
	}
}
